Feature - Enable inspect package `stack()` as acceptable `send_exception` traceback parameter

MVV90 opened the pull request:
As a developer, I want to be able to send my traceback to raygun using something other than a python `traceback` and be able to use something like `inspect.stack()` instead. 

TODO:
- [x] Feature for Python2
- [x] Feature for Python3
- [x] Test coverage

MVV90 commented on python2/raygun4py/raygunmsgs.py:
```diff
@@ -135,7 +135,10 @@ def __init__(self, exc_type, exc_value, exc_traceback, options):
+            if type(exc_traceback) == list and type(exc_traceback[0]) == tuple and str(exc_traceback[0][0].__class__) == "<type 'frame'>":
+                frames = exc_traceback
+            else:
+                frames = inspect.getinnerframes(exc_traceback)
```
From my understanding, `inspect.getinnerframes(exc_traceback)` returns a similar array of frames as `inspect.stack()`. So theoretically, they should be interchangeable without breaking any functionality.

MVV90 commented on .travis.yml:
```diff
@@ -4,7 +4,7 @@ language: python
 cache: pip
 
-python: ["2.7", "3.2", "3.3", "3.4", "3.5", "3.6", "pypy"]
+python: ["2.7", "3.4", "3.5", "3.6", "pypy"]
```
Seems 3.3 and 3.2 are not archives anymore? Only removed to get CI to pass green.
https://github.com/travis-ci/travis-ci/issues/9133#issuecomment-506303002

MVV90 replied:
@UberMouse thoughts?

MVV90 replied:
@samuel-holt ? anybody?

samuel-holt replied:
This seems like quite a common issue. What was the actual error in the CI?

samuel-holt replied:
Have you looked at a workaround like this one: https://github.com/aws/base64io-python/issues/4#issuecomment-412696588?

MVV90 replied:
Cool, I'll check it out.

MVV90 replied:
Got it fixed. The older Ubuntu version happens to have support for Python 3.2 & 3.3.
```
    - python: 3.2
      dist: trusty
```

samuel-holt replied:
Awesome! Good find 👍

MVV90 replied:
Thanks

samuel-holt replied:
Now you've fixed the issue causing CI to fail, can you re-instate versions 3.2 and 3.3?

MVV90 replied:
Yep, confused me too. The `matrix` will run those versions, even though they are not in the `python: []` list.

UberMouse did the [same thing here](https://github.com/MindscapeHQ/raygun4py/commit/2e7149094bc3e850783256b170bc91aa9eda9fe9), he removed 3.7 from the list but also added it to matrix.

I even tried it with both `matrix` and inside `python: []` and it duplicates 3.2 and 3.3 builds. One of the 3.2 builds succeeds, and the other fails.
<img width="1156" alt="Screen Shot 2019-09-04 at 10 50 01 PM" src="https://user-images.githubusercontent.com/14981893/64308367-fffe9900-cf66-11e9-80b4-c58b67dd8dd3.png">

samuel-holt commented on python2/raygun4py/raygunmsgs.py:
```diff
@@ -135,7 +135,10 @@ def __init__(self, exc_type, exc_value, exc_traceback, options):
         frames = None
         try:
-            frames = inspect.getinnerframes(exc_traceback)
+            if type(exc_traceback) == list and type(exc_traceback[0]) == tuple and 'frame' in str(type(exc_traceback[0][0])).lower():
```
This line is a bit hard to understand, could you create a method like `isStack()` which will contain all the conditional logic from this line? It would also be good to write some more tests for this as the current tests only really cover the happy-path. For example, what if `exc_traceback` is an empty list?

It would be good to go a step further and move all this logic into a separate `getFrames()` method.

MVV90 replied:
Yep, good point.

samuel-holt commented on python3/raygun4py/raygunmsgs.py:
```diff
@@ -185,6 +188,16 @@ def __init__(self, exc_type, exc_value, exc_traceback, options):
+            return inspect.getinnerframes(exc_traceback)
+
+    def _is_stack_frame_type(self, exc_traceback):
+        matches = re.findall(self.INSPECT_FRAME_TYPE_SIGNATURE, str(exc_traceback).lower())
```
Is the `findall` method necessary here? Could we use a cheaper method like `match()` instead?

MVV90 commented on python2/raygun4py/raygunmsgs.py:
```diff
@@ -128,14 +129,16 @@ def get_details(self):
 
 class RaygunErrorMessage(object):
 
+    INSPECT_FRAME_TYPE_SIGNATURE = r'<frame [\w ]*0x\w*[>,]'
```
Seems like this regex works relatively well for both Python 2 & 3.

https://regexr.com/

![Screen Shot 2019-09-04 at 5 49 45 PM](https://user-images.githubusercontent.com/14981893/64298406-6bcf0a80-cf43-11e9-8b20-c0e6415a63cd.png)

MVV90 commented:
@samuel-holt When you get a moment to re-review, I think I addressed all your comments.

If there is anything else you'd like me to explain, or fix, I'll try my best.

samuel-holt commented:
@MVV90 Sorry for the delay, I expect these changes will produce the desired result, but I'm not sure the implementation is optimal. I've been looking into a better way to identify whether the parameter passed into the `_get_frames` method matches the result of `inspect.stack()` (and is definitely *only* from this method), but haven't found anything rock solid yet.

I'd be happy to approve this if you could find a better way to distinctly identify that the parameter is from `inspect.stack()` _or_ there were some more tests to cover other potential values which could be passed into this method and validate that the output is still correct.

MVV90 commented on python2/raygun4py/raygunmsgs.py:
```diff
@@ -173,6 +180,28 @@ def __init__(self, exc_type, exc_value, exc_traceback, options):
+                return False
+            else:
+                if not inspect.isframe(frame[0]):
+                    raise DeveloperException("Expected frame type. Got '{}' instead.".format(type(frame[0])))
```
I've seen other projects (outside of raygun) use this style. (`raise DeveloperException`) Not exactly sure if it's out of convention.

My goal is to relate when a developer may have made a mistake by passing an incorrect argument.

but seeing how this is Raygun.. would it be better to send this information to Raygun instead? Silence the python exception and send it to Raygun? Maybe `return None`, which would give the developer a blank traceback, in which case, maybe they'd realize they are using raygun package incorrectly.

```suggestion
                    return None
```

MVV90 commented:
@samuel-holt 

I think I found a better way. After reading through the inspect package docs, they have a `isframe()` method and a `istraceback()` method.
https://docs.python.org/3/library/inspect.html#inspect.isframe

I removed all regex, and used said methods. It's a little easier for python3, as `inspect.stack()` returns a list of `FrameInfo` classes which can easily be compared:
```
import inspect
type(inspect.stack()[0]) == inspect.FrameInfo
```

But python2 is a list of `tuples` which isn't really helpful.. But `isframe()` does work on the actual 'frame' classes in the tuples:
```
import inspect
inspect.isframe(inspect.stack()[0][0])
# True
```

That being said, I added some extra specs/tests for these edge cases. Even trying to spoof the `frames` doesn't work, which is super cool:
```
class frame(object):
    def __init__(self):
        pass

inspect.isframe(frame())
# False
```

Let me know what you think.
